### ISSUE TITLE
Rule idea: `selector-pseudo-class-order`

### ISSUE THREAD

**Issue by evilebottnawi** [stylelint/stylelint#1813]
Given:
```json
{
  "rules": {
    "selector-pseudo-class-order": [["hover", "focus", "not"], { "unspecified": "bottom" }]
  }
}
```

Valid:
```css
.selector:hover,
.selector:focus,
.selector:not(.some),
.selector:active {
  color: #ffffff;
}
```

Invalid:
```css
.selector:focus,
.selector:not(.some) 
.selector:hover {
  color: #ffffff;
}
```

**Comment by jeddy3** [comment]
Sgtm.

**Comment by jeddy3** [comment]
Actually, this assumes a *selector list* of *compound selectors*.

It falls down when their are *complex selectors* and in the *selector list* e.g. 

```
.selector:hover .child:not(.some), 
.selector:active {
  color: red;
}
```

So, on second thoughts, this isn't a clear-cut as our current implementation of `declaration-block-properties-order` (which has one axis), as this has two axis. As such, it no longer meets our rule criteria. 

Any objections if I mark this as a plugin?

**Comment by jeddy3** [comment]
> maybe see only on first `pseudo-class` selector?

That's a decision for the plugin. 

Either way, such a rule isn't "finishable" and should be a plugin.

**Comment by evilebottnawi** [comment]
@jeddy3 How about to look for the order only for identical `selectors`.
Example above not trigger, because we see only for `.selector`. And the rule will be "finishable".

**Comment by evilebottnawi** [comment]
I've seen a few plugins to solve this problem, if there are several plugins, which means that it takes several people and it is logical to create a more general rule to help them.

**Comment by evilebottnawi** [comment]
Eslint ignore the message is full of community and this is not good, because of this there are many plugins, making about the same. I agree that the plugins is good, but should be the "golden section", they have to solve only individual problems, not to expand the core deficiencies. And so I believe that we should study in more detail questions with regards to the rules. Example good plugin `stylelint-number-z-index-constraint` it is only for individuals problems, but `stylelint-selector-pseudo-class-lvhfa` is not good plugin. I'm sure we'll get many more requests for the sorting of `pseudo-elements` and `pseudo-classes`, because it is a good practice, it improves readability and maintainability of code.

**Comment by jeddy3** [comment]
I think I understand what you're saying.

However, I think we've mitigated the problem already by having a three layer system:

1. Core rules.
2. Blessed community plugins that appear on our curated plugin list.
3. Other community plugins.

Having the curated plugin list is key, as it allows us to draw a consistent line between rules and plugins without the fear that classifying something as a plugin is to cast it out into the wild. We link to the plugin list from the very top of the [rules page](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/rules.md#rules) - plugins are very much a part of the stylelint ecosystem.

We cannot control what plugins people write, nor do I think we should. A curated plugin list is our means of providing the best possible user experience while ensuring development on the linter remains sustainable. The latter is a legitimate issue that we faced a few months ago and the three-tier system was put in place to overcome it.

We maintain a good user experience by:

1. Making the best plugins easily findable.
2. Consistently setting expectations about the differing capabilities of core vs plugins.

> which means that it takes several people and it is logical to create a more general rule to help them.

I agree with this. Just in the same way as we replaced the need for rules like `no-extends` and `no-debug` with the more generic `at-rule-whitelist`.

> I'm sure we'll get many more requests for the sorting of `pseudo-elements` and `pseudo-classes`, because it is a good practice, it improves readability and maintainability of code.

Agreed, and is why such a plugin would likely be added to the curated list of plugins.

So, we can suggest, perhaps with your help, that the authors of the various pseudo class ordering plugins pull their efforts together to create a more generic plugin.

**Comment by evilebottnawi** [comment]
@jeddy3 what do you think about check order only on the same `selector`?
We have `["hover", "active", "not"]`.
In example above:
```css
.selector:hover .child:not(.some), 
.selector:active {
  color: red;
}
```
We check order for `.selector` and `.child` separately and all valid.
Also valid:
```css
.selector:hover .child:hover, 
.selector:hover .child:not(.some){
  color: red;
}
```
All algorithm:
1. If we have first some equals selectors (`.selector`) with pseudo classes, check order.
2. If we have second equals selectors (`.child`) with pseudo classes, we check first selectors (`.selector.class`), we check order for equals first selectors.
Valid:
```css
.selector.class .child:hover,
.selector.class .child:active,
.selector.class .child:not(.class) {
  color: red;
}
```
But not trigger errors (because we have difference first selectors):
```css
.selector2.class .child:active,
.selector3.class .child:not(.class),
.selector1.class .child:hover {
  color: red;
}
```
Also valid and working, because `hover` not fire error, we have difference selectors `.selector.class` and `.selector1.class`, and for `.selector.class` order right, `.selector1.class` only one and not fire error.
```css
.selector.class .child:active,
.selector1.class .child:hover
.selector.class .child:not(.class), {
  color: red;
}
```

**Comment by jeddy3** [comment]
> what do you think about check order only on the same selector?

Yes, but as a plugin.

**Comment by evilebottnawi** [comment]
@jeddy3 okey :sob:
